crates/ursa-pod/src/client.rs

```diff
@@ -0,0 +1,85 @@
+use bytes::BytesMut;
+use futures::SinkExt;
+use tokio::io::{AsyncRead, AsyncWrite};
+use tokio_stream::StreamExt;
+use tokio_util::codec::Framed;
+use tracing::{debug, info};
+
+use crate::{
+    codec::{UrsaCodec, UrsaCodecError, UrsaFrame},
+    types::Blake3CID,
+};
+
+pub struct UfdpClient<S: AsyncRead + AsyncWrite + Unpin> {
+    transport: Framed<S, UrsaCodec>,
+}
+
+impl<S> UfdpClient<S>
+where
+    S: AsyncRead + AsyncWrite + Unpin,
+{
+    pub async fn new(stream: S) -> Result<Self, UrsaCodecError> {
+        let codec = UrsaCodec::default();
+
+        let mut transport = Framed::new(stream, codec);
+
+        // send handshake
+        transport
+            .send(UrsaFrame::HandshakeRequest {
+                version: 0,
+                supported_compression_bitmap: 0,
+                lane: 0xFF,
+                pubkey: [1; 48],
+            })
+            .await
+            .expect("handshake request");
+
+        // receive handshake
+        if let Ok(frame) = transport.next().await.expect("handshake response") {
+            match frame {
+                UrsaFrame::HandshakeResponse { .. } => {
+                    info!("received handshake response from server: {frame:?}");
+                }
+                f => return Err(UrsaCodecError::UnexpectedFrame(f.tag().unwrap())),
+            }
+        }
+
+        Ok(Self { transport })
+    }
+
+    pub async fn request(&mut self, hash: Blake3CID) -> Result<BytesMut, UrsaCodecError> {
+        self.transport
+            .send(UrsaFrame::ContentRequest { hash })
+            .await
+            .expect("content request");
+
+        match self.transport.next().await.expect("content response")? {
+            UrsaFrame::ContentResponse {
+                content_len,
+                proof_len,
+                ..
+            } => {
+                info!("received content response");
+
+                debug!("streaming proof ({proof_len})");
+                if proof_len != 0 {
+                    unimplemented!()
+                }
+
+                debug!("streaming content ({content_len})");
+                if content_len != 0 {
+                    self.transport
+                        .codec_mut()
+                        .read_buffer(content_len as usize, 16384);
+                    match self.transport.next().await.expect("content buffer")? {
+                        UrsaFrame::Buffer(bytes) => Ok(bytes),
+                        f => Err(UrsaCodecError::UnexpectedFrame(f.tag().unwrap())),
```

Comment:
Handle the complete read of the response when there are more than 16KiB in the response. I think the `UfdpResponse` should be its own struct implementing `AsyncRead` that reads the content up to completion and handles any verification state.

Although right now you can just return the response bytes as they come in for this iteration, in the future, our intended behavior is for the `UfdpResponse` to only output validated chunks of data.

```rs
impl UfdpClient {
  fn request(&mut self, cid: ...) -> UrsaResponse {
    ..
  }
}

struct UrsaResponse<'client> {
  client: &mut 'client,
  bytes: Bytes
}

impl AsyncRead for UrsaResponse {
..
}

Reply:
implemented `Stream` instead, which can be wrapped with `tokio_util::io::StreamReader` for `AsyncRead`